xarray/core/dataarray.py

```
@@ -1736,5 +1736,93 @@ def dot(self, other):
+                * lower: ``i``.
+                * higher: ``j``.
+                * nearest: ``i`` or ``j``, whichever is nearest.
+                * midpoint: ``(i + j) / 2``.
+
+        Returns
+        -------
+        quantiles : DataArray
+            If `q` is a single quantile and `axis=None`, then the result
+            is a scalar. If multiple percentiles are given, first axis of
+            the result corresponds to the quantile and a quantile dimension
+            is added to the return array. The other axes are the axes that
+            remain after the reduction of the array. If the input
+            contains integers or floats smaller than ``float64``, the output
+            data-type is ``float64``. Otherwise, the output data-type is the
+            same as that of the input.
+        See Also
+        --------
+        np.nanpercentile
+        """
+
+        if dim is not None and axis is not None:
+            raise ValueError("cannot supply both 'axis' and 'dim' arguments")
+
+        isscalar = np.isscalar(q)
+        if isscalar:
+            q = float(q)
+        else:
+            q = np.asarray(q, dtype=np.float64)
+
+        new_dims = list(self.dims)
+        if dim is not None:
+            if isinstance(dim, basestring):
+                axis = self.get_axis_num(dim)
+                new_dims.remove(dim)
+            else:
+                axis = [self.get_axis_num(d) for d in dim]
+                for d in dim:
+                    new_dims.remove(d)
+        elif axis is not None:
+            if hasattr(axis, '__iter__'):
+                for i in axis:
+                    new_dims.remove(self.dims[i])
+            else:
+                new_dims.remove(self.dims[axis])
+        else:
+            new_dims = []
+        # only add the quantile dimension if q is array like
+        if not isscalar:
+            new_dims = ['quantile'] + new_dims
+
+        ps = np.nanpercentile(self.data, q, axis=axis,
```

Let's make sure this fails loudly on dask arrays rather than implicitly converting to numpy.